Trim log messages

a StormCrawler user reported the following problem when processing [http://lijit.com/robots.txt ](http://lijit.com/robots.txt )

the document consists of one very large JSON doc on a single line; the parser obviously complains about its content and adds the content of the line onto the logs; resulting in the error message below.

This could be avoided if we trimmed the log message to the first N characters.


```
2020-01-30 21:44:25,953 FetcherThread #1165 ERROR Unable to write to stream UDP:localhost:514 for appender syslog: org.apache.logging.log4j.core.appender.AppenderLoggingException: Error flushing stream UDP:localhost:514

2020-01-30 21:44:25,954 FetcherThread #1165 ERROR An exception occurred processing Appender syslog org.apache.logging.log4j.core.appender.AppenderLoggingException: Error flushing stream UDP:localhost:514

        at org.apache.logging.log4j.core.appender.OutputStreamManager.flushDestination(OutputStreamManager.java:272)

        at org.apache.logging.log4j.core.appender.OutputStreamManager.write(OutputStreamManager.java:228)

        at org.apache.logging.log4j.core.appender.OutputStreamManager.write(OutputStreamManager.java:201)

        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.writeByteArrayToManager(AbstractOutputStreamAppender.java:186)

        at org.apache.logging.log4j.core.appender.SocketAppender.directEncodeEvent(SocketAppender.java:446)

        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:170)

        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:161)

        at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)

        at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)

        at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)

        at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)

        at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:448)

        at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:433)

        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:417)

        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:403)

        at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)

        at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)

        at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2091)

        at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2005)

        at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1876)

        at org.apache.logging.slf4j.Log4jLogger.warn(Log4jLogger.java:244)

        at crawlercommons.robots.SimpleRobotRulesParser.reportWarning(SimpleRobotRulesParser.java:556)

        at crawlercommons.robots.SimpleRobotRulesParser.parseContent(SimpleRobotRulesParser.java:521)

        at com.digitalpebble.stormcrawler.protocol.RobotRulesParser.parseRules(RobotRulesParser.java:165)

        at com.digitalpebble.stormcrawler.protocol.HttpRobotRulesParser.getRobotRulesSet(HttpRobotRulesParser.java:181)

        at com.digitalpebble.stormcrawler.protocol.RobotRulesParser.getRobotRulesSet(RobotRulesParser.java:175)

        at com.digitalpebble.stormcrawler.protocol.AbstractHttpProtocol.getRobotRules(AbstractHttpProtocol.java:65)

        at com.digitalpebble.stormcrawler.bolt.FetcherBolt$FetcherThread.run(FetcherBolt.java:459)

Caused by: java.io.IOException: Message too long (sendto failed)

        at java.net.PlainDatagramSocketImpl.send(Native Method)

        at java.net.DatagramSocket.send(DatagramSocket.java:693)

        at org.apache.logging.log4j.core.net.DatagramOutputStream.flush(DatagramOutputStream.java:103)

        at org.apache.logging.log4j.core.appender.OutputStreamManager.flushDestination(OutputStreamManager.java:270)

        ... 27 more
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,6 +1,7 @@
 Crawler-Commons Change Log
 
 Current Development 1.1-SNAPSHOT (yyyy-mm-dd)
+- SimpleRobotRulesParser: Trim log messages (jnioche, sebastian-nagel) #281
 - SimpleRobotRulesParser: counter _numWarnings not thread-safe (sebastian-nagel, kkrugler) #278
 - ParameterizedTest not executed by mvn builds (sebastian-nagel) #273
 - [BasicNormalizer] Empty path before query to be normalized to `/` (Chaiavi, sebastian-nagel) #247

--- a/src/main/java/crawlercommons/robots/SimpleRobotRulesParser.java
+++ b/src/main/java/crawlercommons/robots/SimpleRobotRulesParser.java
@@ -561,6 +561,12 @@ public class SimpleRobotRulesParser extends BaseRobotsParser {
         }
 
         if (state._numWarnings < _maxWarnings) {
+            for (int i = 0; i < args.length; i++) {
+                if (args[i] instanceof String && ((String) args[i]).length() > 1024) {
+                    // clip overlong strings to prevent from overflows in log messages
+                    args[i] = ((String) args[i]).substring(0, 1024) + " ...";
+                }
+            }
             LOGGER.warn("\t " + msg, args);
         }
     }

